test/src/test/java/jenkins/security/BasicHeaderProcessorTest.java

```diff
@@ -155,4 +188,79 @@ public HttpResponse doIndex() {
             return HttpResponses.plainText(u!=null ? u.getId() : "anonymous");
         }
     }
+    
+    @TestExtension
+    public static class DummySecurityListener extends SecurityListener {
+        final EventQueue<UserDetails> authenticatedCalls;
+        final EventQueue<String> failedToAuthenticateCalls;
+        final EventQueue<String> loggedInCalls;
+
+        public DummySecurityListener(){
+            this.authenticatedCalls = new EventQueue<>();
+            this.failedToAuthenticateCalls = new EventQueue<>();
+            this.loggedInCalls = new EventQueue<>();
+        }
+        
+        void clearPreviousCalls(){
+            this.authenticatedCalls.clear();
+            this.failedToAuthenticateCalls.clear();
+            this.loggedInCalls.clear();
+        }
+        
+        @Override
+        protected void authenticated(@Nonnull UserDetails details) {
+            this.authenticatedCalls.add(details);
+        }
+
+        @Override
+        protected void failedToAuthenticate(@Nonnull String username) {
+            this.failedToAuthenticateCalls.add(username);
+        }
+
+        @Override
+        protected void loggedIn(@Nonnull String username) {
+            this.loggedInCalls.add(username);
+        }
+
+        @Override
+        protected void failedToLogIn(@Nonnull String username) {
+        
+        }
+    
+        @Override
+        protected void loggedOut(@Nonnull String username) {
+        
+        }
+    }
+    
+    private static class EventQueue<T> {
+        private final List<T> eventList = new ArrayList<>();
+    
+        void assertLastEventIs(T expected){
+            assertLastEventIs(actual -> actual.equals(expected));
+        }
+
+        void assertLastEventIs(Predicate<T> predicate){
+            if(eventList.isEmpty()){
+                fail("event list is empty");
+            }
+        
+            T t = eventList.remove(eventList.size() - 1);
+            assertTrue(predicate.test(t));
+            eventList.clear();
```

Comment:
Self-modifying asserts is something unexpected to me. I would advice to document or rework it if it gets moved to JTH

Reply:
Method renamed to explicitly say we remove the event after the assert